### PR TITLE
Enh/improve error handling

### PR DESCRIPTION
**Pull request by leiter-jakab**
Hi IPFS Team, great project. Hoping to help with this change. Not sure if this is the correct solution, but I would love to hear your feedback.

Regards,
Gergo

### CODE REVIEW

File: fslock.go [hunk 1]
```diff
@@ -22,7 +23,30 @@ func errPerm(path string) error {
 
 // Lock creates the lock.
 func Lock(confdir, lockFile string) (io.Closer, error) {
-	return lock.Lock(filepath.Join(confdir, lockFile))
+	lk, err := lock.Lock(filepath.Join(confdir, lockFile))
+	if err != nil {
+		// EAGAIN == someone else has the lock
+		if err == syscall.EAGAIN {
+			return lk, errors.New(fmt.Sprintf("Someone else has the lock: %s", filepath.Join(confdir, lockFile)))
```
Comment (Stebalien) [inline fslock.go]:
```suggestion
			return lk, fmt.Errorf("Someone else has the lock: %s", filepath.Join(confdir, lockFile))
```

File: fslock.go [hunk 2]
```diff
@@ -35,30 +59,20 @@ func Locked(confdir, lockFile string) (bool, error) {
 
 	lk, err := Lock(confdir, lockFile)
 	if err != nil {
```
Comment (Stebalien) [inline fslock.go]: Given that we're now parsing the error in `Lock`, what if we just had `Lock` return a custom error type on failure? We could return an error implementing:

```go
type myerror interface {
  Permission() bool // returns true if this is a permission error.
  Locked() bool // returns true if this error means that the lock has already been taken.
}
```

Then we can provide two functions to determine why taking the lock failed:

```go
func IsLocked(err) bool { ... }
func IsPermission(err) bool { ... }
```

_This_ function would be reduced to an `IsLocked(err)` call (and a debug statement).

Comment (leiter-jakab) [reply fslock.go]: @Stebalien I have been thinking about this and I have to learn more about idiomatic error handling in Go before I make any changes.

On the other hand, I'm not sure that the [original issue](https://github.com/ipfs/go-ipfs/issues/6434) has been solved by my changes.

What do you think?

Comment (Stebalien) [reply fslock.go]: The original issue has been solved, I'm just trying to avoid parsing errors in two places.

File: fslock.go [hunk 3]
```diff
@@ -35,30 +58,20 @@ func Locked(confdir, lockFile string) (bool, error) {
...
+		if strings.Contains(errCase, "Lock is already held by us:") {
+			log.Debugf(errCase)
 			return true, nil
 		}
-
-		// we hold the lock ourselves
-		if strings.Contains(err.Error(), "already locked") {
-			log.Debugf("Lock is already held by us: %s", filepath.Join(confdir, lockFile))
+		if strings.Contains(errCase, "Someone else has the lock:") {
+			log.Debugf(errCase)
```
Comment (Stebalien) [inline fslock.go]: This should be `log.Debug(errCase)` (otherwise, we'll parse `errCase` for string format specifiers).

File: fslock.go [hunk 4]
```diff
@@ -35,30 +58,20 @@ func Locked(confdir, lockFile string) (bool, error) {
...
 		// lock fails on permissions error
-		if os.IsPermission(err) {
-			log.Debugf("Lock fails on permissions error")
-			return false, errPerm(confdir)
-		}
-		if isLockCreatePermFail(err) {
-			log.Debugf("Lock fails on permissions error")
-			return false, errPerm(confdir)
+		if strings.Contains(errCase, "permissions denied") {
+			log.Debugf(errCase)
```
Comment (Stebalien) [inline fslock.go]: ditto.

File: fslock.go [hunk 5]
```diff
@@ -35,30 +58,20 @@ func Locked(confdir, lockFile string) (bool, error) {
...
-		// EAGAIN == someone else has the lock
-		if err == syscall.EAGAIN {
-			log.Debugf("Someone else has the lock: %s", filepath.Join(confdir, lockFile))
-			return true, nil
-		}
-		if strings.Contains(err.Error(), "resource temporarily unavailable") {
-			log.Debugf("Can't lock file: %s.\n reason: %s", filepath.Join(confdir, lockFile), err.Error())
+		errCase := err.Error()
+		if strings.Contains(errCase, "Lock is already held by us:") {
+			log.Debugf(errCase)
```
Comment (Stebalien) [inline fslock.go]: ditto

**Comment by Stebalien** [comment]
I went digging through some go error documentation and realized we could use the built-in error types. I took the liberty of making those changes rather than try to change direction on you again.

**Comment by leiter-jakab** [comment]
> I went digging through some go error documentation and realized we could use the built-in error types. I took the liberty of making those changes rather than try to change direction on you again.

Very nice, I learned quite a bit. Could you please point me to the documentation where you found the relevant info? Thanks.

**Comment by Stebalien** [comment]
* As: https://golang.org/pkg/errors/#example_As
* PathError: https://golang.org/pkg/os/#PathError
* Definition of IsPermission: https://golang.org/src/os/error.go?s=3108:3141#L89

**Pull request merged**